[Bug report] Flink MULTISET type causes schema mismatch error when writing to Paimon table via Gravitino Flink Connector

### Version

main branch

### Describe what's wrong

When a Paimon table contains a column with `MULTISET<STRING>` type, writing to the table through the Gravitino Flink Connector fails with a schema mismatch error. The same operation works correctly when using the native Paimon Flink Catalog directly (without Gravitino).

Root cause: Gravitino's type system does not have a `MULTISET` type. In the Paimon catalog backend, `MULTISET<T>` is converted to `MapType<T, INTEGER>` (in `PaimonToGravitinoTypeVisitor.visit(MultisetType)`). When the Flink Connector reads the table back, it converts this `MapType` to Flink's `MAP<T, INT NOT NULL>` instead of restoring it to `MULTISET<T>`. Paimon's `FlinkTableFactory` then compares the Flink schema (`MAP<STRING, INT NOT NULL>`) against the store schema (`MULTISET<STRING>`) and throws an `IllegalArgumentException`.

Additionally, `TypeUtils.toGravitinoType()` in the Flink Connector throws `UnsupportedOperationException` for `MULTISET` type, which also blocks table creation with MULTISET columns.

### Error message and/or stacktrace

Caused by: java.lang.IllegalArgumentException: Flink schema and store schema are not the same,
store schema is ROW<..., `field_multiset` MULTISET<STRING> COMMENT '测试字段', ...>,
Flink schema is ROW<..., `field_multiset` MAP<STRING, INT NOT NULL> '...', ...> NOT NULL
    at org.apache.paimon.utils.Preconditions.checkArgument(Preconditions.java:149)
    at org.apache.paimon.flink.AbstractFlinkTableFactory.buildPaimonTable(AbstractFlinkTableFactory.java:246)
    at org.apache.paimon.flink.AbstractFlinkTableFactory.createDynamicTableSink(AbstractFlinkTableFactory.java:127)
    at org.apache.paimon.flink.FlinkTableFactory.createDynamicTableSink(FlinkTableFactory.java:66)
    at org.apache.flink.table.factories.FactoryUtil.createDynamicTableSink(FactoryUtil.java:335)
    at org.apache.flink.table.planner.delegation.PlannerBase.getTableSink(PlannerBase.scala:450)
    at org.apache.flink.table.planner.delegation.PlannerBase.translateToRel(PlannerBase.scala:227)


### How to reproduce

1. Create a Paimon table with a `MULTISET<STRING>` column (either directly via Paimon or via Flink SQL without Gravitino):

```sql
CREATE TABLE test_multiset (
    id BIGINT,
    field_multiset MULTISET<STRING>,
    PRIMARY KEY (id) NOT ENFORCED
);
```

1. Register the Paimon catalog in Gravitino
2. Use the Gravitino Flink Connector (GravitinoCatalogStore) to access the table
3. Attempt to write data to the table:

```sql
INSERT INTO test_multiset VALUES (1, MULTISET['a', 'b', 'a']);
```

1. The INSERT fails with the schema mismatch error above

Environment: Flink 1.20.3, Paimon 1.3.1, Gravitino main branch

### Additional context

The type conversion chain that causes this issue:
Paimon store: MULTISET<STRING>
  → PaimonToGravitinoTypeVisitor.visit(MultisetType): MapType(STRING, INTEGER, valueNullable=false)
  → Flink Connector TypeUtils.toFlinkType(MAP): MAP<STRING, INT NOT NULL>
  → Paimon FlinkTableFactory compares: MAP<STRING, INT NOT NULL> ≠ MULTISET<STRING> → ERROR

--- a/catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TypeUtils.java
+++ b/catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/TypeUtils.java
@@ -28,6 +28,7 @@ import org.apache.paimon.types.BooleanType;
 import org.apache.paimon.types.CharType;
 import org.apache.paimon.types.DataType;
 import org.apache.paimon.types.DataTypeDefaultVisitor;
+import org.apache.paimon.types.DataTypeJsonParser;
 import org.apache.paimon.types.DataTypes;
 import org.apache.paimon.types.DateType;
 import org.apache.paimon.types.DecimalType;
@@ -192,14 +193,10 @@ public class TypeUtils {
 
     @Override
     public Type visit(MultisetType multisetType) {
-      // Unlike a Java Set, MultisetType allows for multiple instances for each of its
-      // elements with a common subtype. And a conversion is possible through a map
-      // that assigns each value to an integer to represent the multiplicity of the values.
-      // For example, a `MULTISET<INT>` is converted to a `MAP<Integer, Integer>`, the key of the
-      // map represents the elements of the Multiset and the value represents the multiplicity of
-      // the elements in the Multiset.
-      return Types.MapType.of(
-          multisetType.getElementType().accept(this), Types.IntegerType.get(), false);
+      // Gravitino's type system does not have a native MULTISET type. We use ExternalType to
+      // preserve the original Paimon MULTISET type information, so it can be correctly restored
+      // when converting back to Paimon or Flink types.
+      return Types.ExternalType.of(multisetType.asSQLString());
     }
 
     @Override
@@ -280,6 +277,18 @@ public class TypeUtils {
                     builder.field(field.name(), dataTypeWithNullable, field.comment());
                   });
           return builder.build();
+        case EXTERNAL:
+          Types.ExternalType externalType = (Types.ExternalType) type;
+          String catalogString = externalType.catalogString();
+          if (catalogString.startsWith("MULTISET<") && catalogString.endsWith(">")) {
+            // Restore MULTISET type: extract element type SQL string and parse it
+            String elementTypeSql =
+                catalogString.substring("MULTISET<".length(), catalogString.length() - 1);
+            DataType elementDataType = DataTypeJsonParser.parseAtomicTypeSQLString(elementTypeSql);
+            return new MultisetType(elementDataType);
+          }
+          throw new UnsupportedOperationException(
+              String.format("Paimon does not support Gravitino external type: %s", catalogString));
         default:
           throw new UnsupportedOperationException(
               String.format(

--- a/flink-connector/flink/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java
+++ b/flink-connector/flink/src/main/java/org/apache/gravitino/flink/connector/utils/TypeUtils.java
@@ -30,7 +30,10 @@ import org.apache.flink.table.types.logical.CharType;
 import org.apache.flink.table.types.logical.DecimalType;
 import org.apache.flink.table.types.logical.LogicalType;
 import org.apache.flink.table.types.logical.MapType;
+import org.apache.flink.table.types.logical.MultisetType;
 import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.table.types.logical.utils.LogicalTypeParser;
+import org.apache.flink.table.types.utils.TypeConversions;
 import org.apache.gravitino.rel.types.Type;
 import org.apache.gravitino.rel.types.Types;
 
@@ -130,6 +133,10 @@ public class TypeUtils {
       case NULL:
         return Types.NullType.get();
       case MULTISET:
+        MultisetType multisetType = (MultisetType) logicalType;
+        // Gravitino's type system does not have a native MULTISET type. We use ExternalType to
+        // preserve the original type information so it can be correctly restored.
+        return Types.ExternalType.of(multisetType.asSerializableString());
       case STRUCTURED_TYPE:
       case UNRESOLVED:
       case DISTINCT_TYPE:
@@ -236,6 +243,14 @@ public class TypeUtils {
         return DataTypes.INTERVAL(DataTypes.YEAR());
       case INTERVAL_DAY:
         return DataTypes.INTERVAL(DataTypes.DAY());
+      case EXTERNAL:
+        Types.ExternalType externalType = (Types.ExternalType) gravitinoType;
+        String catalogString = externalType.catalogString();
+        // Parse the external catalog type string back to Flink LogicalType.
+        // This is used to restore types like MULTISET that Gravitino doesn't natively support.
+        LogicalType parsedType =
+            LogicalTypeParser.parse(catalogString, TypeUtils.class.getClassLoader());
+        return TypeConversions.fromLogicalToDataType(parsedType);
       default:
         throw new UnsupportedOperationException("Not support " + gravitinoType.toString());
     }

--- a/flink-connector/flink/src/test/java/org/apache/gravitino/flink/connector/integration/test/paimon/FlinkPaimonCatalogIT.java
+++ b/flink-connector/flink/src/test/java/org/apache/gravitino/flink/connector/integration/test/paimon/FlinkPaimonCatalogIT.java
@@ -18,18 +18,29 @@
  */
 package org.apache.gravitino.flink.connector.integration.test.paimon;
 
+import static org.apache.gravitino.flink.connector.integration.test.utils.TestUtils.toFlinkPhysicalColumn;
+import static org.junit.jupiter.api.Assertions.fail;
+
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Lists;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import org.apache.flink.table.api.DataTypes;
 import org.apache.flink.table.api.TableResult;
+import org.apache.flink.table.catalog.CatalogBaseTable;
+import org.apache.flink.table.catalog.ObjectPath;
+import org.apache.flink.table.catalog.exceptions.TableNotExistException;
 import org.apache.flink.types.Row;
 import org.apache.gravitino.Catalog;
 import org.apache.gravitino.NameIdentifier;
 import org.apache.gravitino.flink.connector.integration.test.FlinkCommonIT;
+import org.apache.gravitino.rel.Column;
 import org.apache.gravitino.rel.Table;
 import org.apache.gravitino.rel.expressions.distributions.Distributions;
 import org.apache.gravitino.rel.expressions.distributions.Strategy;
+import org.apache.gravitino.rel.types.Types;
 import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.BeforeAll;
@@ -223,5 +234,59 @@ public abstract class FlinkPaimonCatalogIT extends FlinkCommonIT {
     Assertions.assertFalse(metalake.catalogExists(catalogName));
     Assertions.assertEquals(
         numCatalogs, tableEnv.listCatalogs().length, "The created catalog should be dropped.");
+  }
+
+  @Test
+  public void testMultisetType() {
+    String databaseName = "test_multiset_type_db";
+    String tableName = "test_multiset_table";
+
+    doWithSchema(
+        currentCatalog(),
+        databaseName,
+        catalog -> {
+          // Create a table with MULTISET column using ExternalType through Gravitino API.
+          // This simulates the scenario where a Paimon table with MULTISET type is registered
+          // in Gravitino (e.g., created by native Paimon and then managed by Gravitino).
+          Column[] columns =
+              new Column[] {
+                Column.of("id", Types.LongType.get(), "id"),
+                Column.of(
+                    "field_multiset", Types.ExternalType.of("MULTISET<STRING>"), "multiset field")
+              };
+          catalog
+              .asTableCatalog()
+              .createTable(
+                  NameIdentifier.of(databaseName, tableName),
+                  columns,
+                  "test multiset table",
+                  ImmutableMap.of());
+
+          // Verify that getTable through Gravitino Flink Connector returns MULTISET type
+          Optional<org.apache.flink.table.catalog.Catalog> flinkCatalog =
+              tableEnv.getCatalog(catalog.name());
+          Assertions.assertTrue(flinkCatalog.isPresent());
+          try {
+            CatalogBaseTable table =
+                flinkCatalog.get().getTable(new ObjectPath(databaseName, tableName));
+            Assertions.assertNotNull(table);
+
+            org.apache.flink.table.catalog.Column[] expected =
+                new org.apache.flink.table.catalog.Column[] {
+                  org.apache.flink.table.catalog.Column.physical("id", DataTypes.BIGINT())
+                      .withComment("id"),
+                  org.apache.flink.table.catalog.Column.physical(
+                          "field_multiset", DataTypes.MULTISET(DataTypes.STRING()))
+                      .withComment("multiset field")
+                };
+            org.apache.flink.table.catalog.Column[] actual =
+                toFlinkPhysicalColumn(table.getUnresolvedSchema().getColumns());
+            Assertions.assertArrayEquals(expected, actual);
+          } catch (TableNotExistException e) {
+            fail(e);
+          }
+        },
+        true,
+        supportDropCascade());
   }
 }

--- a/flink-connector/flink/src/test/java/org/apache/gravitino/flink/connector/utils/TestTypeUtils.java
+++ b/flink-connector/flink/src/test/java/org/apache/gravitino/flink/connector/utils/TestTypeUtils.java
@@ -33,6 +33,7 @@ import org.apache.flink.table.types.logical.DoubleType;
 import org.apache.flink.table.types.logical.IntType;
 import org.apache.flink.table.types.logical.LocalZonedTimestampType;
 import org.apache.flink.table.types.logical.MapType;
+import org.apache.flink.table.types.logical.MultisetType;
 import org.apache.flink.table.types.logical.NullType;
 import org.apache.flink.table.types.logical.RowType;
 import org.apache.flink.table.types.logical.SmallIntType;
@@ -252,5 +253,36 @@ public class TestTypeUtils {
     Assertions.assertThrows(
         UnsupportedOperationException.class,
         () -> TypeUtils.toFlinkType(Types.TimestampType.withTimeZone(10)));
+  }
+
+  @Test
+  public void testMultisetTypeConversion() {
+    // MULTISET<STRING> (VARCHAR(MAX)) should be preserved as ExternalType
+    Assertions.assertEquals(
+        Types.ExternalType.of("MULTISET<VARCHAR(2147483647)>"),
+        TypeUtils.toGravitinoType(new MultisetType(new VarCharType(Integer.MAX_VALUE))));
+
+    // MULTISET<INT> should be preserved as ExternalType
+    Assertions.assertEquals(
+        Types.ExternalType.of("MULTISET<INT>"),
+        TypeUtils.toGravitinoType(new MultisetType(new IntType())));
+
+    // MULTISET<BIGINT> should be preserved as ExternalType
+    Assertions.assertEquals(
+        Types.ExternalType.of("MULTISET<BIGINT>"),
+        TypeUtils.toGravitinoType(new MultisetType(new BigIntType())));
+
+    // ExternalType with MULTISET should be converted back to Flink MULTISET
+    Assertions.assertEquals(
+        DataTypes.MULTISET(DataTypes.STRING()),
+        TypeUtils.toFlinkType(Types.ExternalType.of("MULTISET<VARCHAR(2147483647)>")));
+
+    Assertions.assertEquals(
+        DataTypes.MULTISET(DataTypes.INT()),
+        TypeUtils.toFlinkType(Types.ExternalType.of("MULTISET<INT>")));
+
+    Assertions.assertEquals(
+        DataTypes.MULTISET(DataTypes.BIGINT()),
+        TypeUtils.toFlinkType(Types.ExternalType.of("MULTISET<BIGINT>")));
   }
 }

